contrib/opensearch-project/opensearch-go.v4/opensearch.go

```diff
@@ -0,0 +1,222 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025-present Datadog, Inc.
+
+// Package opensearch provides tracing functions for tracing the opensearch-project/opensearch-go/v4 package (https://github.com/opensearch-project/opensearch-go).
+package opensearch
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+
+	"github.com/opensearch-project/opensearch-go/v4"
+	"github.com/opensearch-project/opensearch-go/v4/opensearchtransport"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/namingschema"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/telemetry"
+)
+
+const (
+	componentName      = "opensearch-project/opensearch-go/v4"
+	defaultServiceName = "opensearch.client"
+)
+
+var (
+	// bodyCutoff specifies the maximum number of bytes that will be stored as a tag
+	// value obtained from an HTTP request or response body.
+	bodyCutoff = 5 * 1024
+
+	_ opensearchtransport.Interface    = (*transport)(nil)
+	_ opensearchtransport.Discoverable = (*transport)(nil)
+	_ http.RoundTripper                = (*roundTripper)(nil)
+)
+
+func init() {
+	telemetry.LoadIntegration(componentName)
+	tracer.MarkIntegrationImported("github.com/opensearch-project/opensearch-go/v4")
+}
+
+// WrapClient wraps OpenSearch client with tracing.
+func WrapClient(c *opensearch.Client, opts ...Option) {
+	tracerConfig := defaultConfig()
+	for _, fn := range opts {
+		fn(tracerConfig)
+	}
+	opensearchtransport := c.Transport
+	t := &transport{
+		opensearchtransport: opensearchtransport,
+		config:              tracerConfig,
+	}
+	c.Transport = t
+}
+
+// NewDefaultClient returns a new default opensearch.Client enhanced with tracing.
+func NewDefaultClient(opts ...Option) (*opensearch.Client, error) {
+	return NewClient(opensearch.Config{}, opts...)
+}
+
+// NewClient returns a new opensearch.Client enhanced with tracing.
+func NewClient(cfg opensearch.Config, opts ...Option) (*opensearch.Client, error) {
+	if cfg.Transport == nil {
+		cfg.Transport = WrapRoundTripper(http.DefaultTransport)
+	} else {
+		cfg.Transport = WrapRoundTripper(cfg.Transport)
+	}
+	c, err := opensearch.NewClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+	tracerConfig := defaultConfig()
+	for _, fn := range opts {
+		fn(tracerConfig)
+	}
+	t := &transport{
+		opensearchtransport: c.Transport,
+		config:              tracerConfig,
+	}
+	c.Transport = t
+	return c, nil
```

Comment:
What is the difference / advantage of tracing both the transport and the roundtripper? Would it be possible to achieve the same functionality using one of the 2 only?